src/memory.rs

```diff
@@ -54,100 +54,138 @@ impl ByteSliceView {
     }
 }
 
-#[no_mangle]
-pub extern "C" fn allocate_rust(ptr: *const u8, length: usize) -> Buffer {
-    // Go doesn't store empty buffers the same way Rust stores empty slices (with NonNull  pointers
-    // equal to the offset of the type, which would be equal to 1 in this case)
-    // so when it wants to represent an empty buffer, it passes a null pointer with 0 length here.
-    if length == 0 {
-        Buffer::from_vec(Vec::new())
-    } else {
-        Buffer::from_vec(Vec::from(unsafe { slice::from_raw_parts(ptr, length) }))
-    }
+/// A view into a `Option<&[u8]>`, created and maintained by Rust.
+///
+/// This can be copied into a []byte in Go.
+#[repr(C)]
+pub struct U8SliceView {
+    /// True if and only if this is None/nil. If this is true, the other fields must be ignored.
+    is_nil: bool,
+    ptr: *const u8,
+    len: usize,
 }
 
-// this frees memory we released earlier
-#[no_mangle]
-pub extern "C" fn free_rust(buf: Buffer) {
-    unsafe {
-        let _ = buf.consume();
+impl U8SliceView {
+    pub fn new(source: Option<&[u8]>) -> Self {
+        match source {
+            Some(data) => Self {
+                is_nil: false,
+                ptr: data.as_ptr(),
+                len: data.len(),
+            },
+            None => Self {
+                is_nil: true,
+                ptr: std::ptr::null::<u8>(),
+                len: 0,
+            },
+        }
     }
 }
 
-#[derive(Copy, Clone)]
+/// A Vector type that requires explicit creation and destruction and
+/// can be sent via FFI.
+/// It can be created from `Option<Vec<u8>>` and be converted into `Option<Vec<u8>>`.
+/// This type is always created in Rust and always dropped in Rust.
```

Comment:
Add: "often in different FFI calls".
Creation and copy from Go happens in one, it is passed in a future call to be consumed.

This "always created in Rust" is true, but the control flow is from Go. This was confusing (also the naming of the call). Maybe better doc strings?